File: frontend/src/features/plan/subthemaweghaling.ts
import { t } from "../../i18n";
import type { Subthemaweghaling } from "../../lib/queries";

/**
 * The sentences the question before taking a subthema out of the agenda says (FB-096), in the order a teacher reads
 * them: which days, which activiteiten go with it, which hoekverrijkingen, and what it does to the dekking.
 *
 * **Each sentence asserts only what the server's counts guarantee.** The dekking sentence is left out for a run drawn
 * from its activiteiten alone: only a stored window makes the goals count (ADR-0047), so removing none changes
 * nothing there. With another window of the subthema staying, it says the goals go on counting instead.
 */
export function weghaalzinnen(
  reeks: { subthemaNaam: string; van: string; tot: string },
  gevolg: Subthemaweghaling,
  dagnaam: (datum: string) => string,
): string[] {
  const zinnen = [
    reeks.van === reeks.tot
      ? t("subthemaWeg.dag", { naam: reeks.subthemaNaam, dag: dagnaam(reeks.van) })
      : t("subthemaWeg.dagen", { naam: reeks.subthemaNaam, van: dagnaam(reeks.van), tot: dagnaam(reeks.tot) }),
  ];

  zinnen.push(
    gevolg.aantalActiviteiten === 0
      ? t("subthemaWeg.geenActiviteiten")
      : gevolg.aantalActiviteiten === 1
        ? t("subthemaWeg.activiteitenEen")
        : t("subthemaWeg.activiteiten", { aantal: gevolg.aantalActiviteiten }),
  );

  if (gevolg.aantalHoekverrijkingen === 1) zinnen.push(t("subthemaWeg.verrijkingenEen"));
  else if (gevolg.aantalHoekverrijkingen > 1) {
    zinnen.push(t("subthemaWeg.verrijkingen", { aantal: gevolg.aantalHoekverrijkingen }));
  }

  if (gevolg.heeftPeriode) {
    zinnen.push(t(gevolg.blijftElders ? "subthemaWeg.dekkingBlijft" : "subthemaWeg.dekkingWeg"));
  }

  zinnen.push(t("algemeen.nietTerugTeDraaien"));
  return zinnen;
}

File: backend/src/Jaarplanner.Application/Planning/Weekplanning/Subthemaweghaling.cs
namespace Jaarplanner.Application.Planning.Weekplanning;

/// <summary>
/// What taking a subthema out of the agenda over a stretch of days would take with it (FB-096): the counts the
/// confirmation names before anything is gone.
/// <para>
/// <b>Computed by the same selection the delete uses</b> (<c>WeekplanningService.SelecteerSubthema</c>), so the numbers
/// a teacher confirms are the rows that go.
/// </para>
/// </summary>
/// <param name="AantalActiviteiten">
/// The activiteitplaatsingen of this subthema on those days, open proposals of a weekvoorstel included: they leave the
/// agenda too.
/// </param>
/// <param name="AantalHoekverrijkingen">
/// The hoekverrijkingen written against the windows that go (FB-020). They cascade with their window, and the owner's
/// rule for that is "mee weg, met aantal" (2026-09-15).
/// </param>
/// <param name="HeeftPeriode">
/// Whether a stored window goes. Only a window makes a subthema's goals count for the klas's dekking (ADR-0047), so a
/// run drawn from its activiteiten alone changes nothing there and the confirmation must not say it does.
/// </param>
/// <param name="BlijftElders">
/// Whether another window of the same subthema stays in this klas's plan, which keeps its goals counting.
/// </param>
public sealed record Subthemaweghaling(
    int AantalActiviteiten,
    int AantalHoekverrijkingen,
    bool HeeftPeriode,
    bool BlijftElders);
